docs/Toy/toy/parser.py

```diff
@@ -0,0 +1,431 @@
+from pathlib import Path
+from typing import List, NoReturn, TypeVar, cast
+
+from toy.lexer import (OperatorToken, Token, tokenize, NumberToken, EOFToken,
+                       IdentifierToken)
+from toy.ast import (BinaryExprAST, ExprAST, NumberExprAST, FunctionAST,
+                     ModuleAST, LiteralExprAST, VarDeclExprAST, VarType,
+                     VariableExprAST, ReturnExprAST, PrintExprAST, CallExprAST,
+                     PrototypeAST)
+
+
+class ParseError(Exception):
+
+    def __init__(self,
+                 token: Token,
+                 expected: str | type[Token],
+                 context: str = '',
+                 line: str = ''):
+        loc = token.loc
+
+        message = f'Parse error ({loc.line}, {loc.col}): expected '
+        if isinstance(expected, str):
+            message += expected
+        else:
+            message += expected.name()
+
+        if len(context):
+            message += ' ' + context
+
+        message += f" but has Token '{token.text}'\n"
+
+        if len(line):
+            message += line + '\n'
+
+        super().__init__(message)
+
+    pass
+
+
+TokenT = TypeVar('TokenT', bound=Token)
+
+
+class Parser:
+    file: Path
+    program: str
+    tokens: List[Token]
+    pos: int
+
+    def __init__(self, file: Path, program: str):
+        self.file = file
+        self.program = program
+        self.tokens = tokenize(file, program)
+        self.pos = 0
+
+    def getToken(self):
```

Comment:
Can we document the few next functions?

Reply:
- [ ] add documentation for `getToken` and token precedence functions